// sanity-plugin-datatable/build/schema/table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _container = _interopRequireDefault(require("../container"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = {
  title: 'Table',
  name: 'table',
  type: 'object',
  fields: [{
    type: 'array',
    name: 'rows',
    of: [{
      type: 'column'
    }]
  }],
  inputComponent: _container["default"],
  preview: {
    select: {
      rows: 'rows'
    },
    prepare: function prepare(_ref) {
      var rows = _ref.rows;
      var rowLength = rows.length || 0;
      var colLength = rowLength ? rows[0].cells.length : 0;

      if (rowLength === 0) {
        return {
          title: 'Empty table',
          subtitle: ''
        };
      }

      var headings = rows[0].cells;
      return {
        title: headings.join(' | '),
        subtitle: "".concat(colLength, "\xA0cols x ").concat(rowLength, " rows")
      };
    }
  }
};
exports["default"] = _default;
