// sanity-plugin-datatable/build/table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _lodash = require("lodash");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _primitives = require("./primitives");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Table = function Table(props) {
  var data = props.data,
      rows = props.rows,
      columns = props.columns,
      handleChange = props.handleChange,
      onRemoveRow = props.onRemoveRow,
      onRemoveColumn = props.onRemoveColumn,
      onReset = props.onReset;
  /**
  * Render rows
  */

  var rowRange = (0, _lodash.range)(rows);
  var rowsList = rowRange.map(function (rowIndex) {
    return /*#__PURE__*/_react["default"].createElement(_primitives.Row, {
      key: rowIndex,
      cols: columns
    }, (0, _lodash.range)(columns).map(function (columnIndex) {
      return /*#__PURE__*/_react["default"].createElement(_primitives.Cell, {
        key: columnIndex,
        type: "text",
        value: (data[rowIndex] || [])[columnIndex] || '',
        onChange: function onChange(e) {
          return handleChange(e, rowIndex, columnIndex);
        }
      });
    }), /*#__PURE__*/_react["default"].createElement(_default2["default"], {
      kind: "simple",
      onClick: function onClick() {
        return onRemoveRow(rowIndex);
      },
      title: "Remove this row"
    }, "-"));
  });
  /**
   * Render
   */

  return /*#__PURE__*/_react["default"].createElement(_primitives.Container, null, rows === 0 && /*#__PURE__*/_react["default"].createElement(_primitives.Empty, {
    children: "Empty table"
  }), rowsList, rows > 0 && /*#__PURE__*/_react["default"].createElement(_primitives.Row, {
    cols: columns
  }, (0, _lodash.range)(columns).map(function (columnIndex) {
    return /*#__PURE__*/_react["default"].createElement(_default2["default"], {
      style: {
        width: '100%'
      },
      key: columnIndex,
      kind: "simple",
      title: "Remove this column",
      onClick: function onClick() {
        return onRemoveColumn(columnIndex);
      }
    }, "-");
  }), /*#__PURE__*/_react["default"].createElement(_default2["default"], {
    color: "danger",
    onClick: onReset,
    title: "Remove all rows and columns"
  }, "\xD7")));
};

Table.propTypes = {
  data: _propTypes["default"].array,
  rows: _propTypes["default"].number,
  columns: _propTypes["default"].number,
  handleChange: _propTypes["default"].func
};
var _default = Table;
exports["default"] = _default;
